### PR TITLE
feat: Use linters from regular dependencies

### PR DESCRIPTION
**Pull request by dickermoshe**
As of now custom_lint doesn't look for linters installed as regular dependencies. This would change that.

Semi-Fix for broken `CustomLintWorkspace resolvePluginHost runs offline if possible` test too

https://github.com/invertase/dart_custom_lint/issues/275

@rrousselGit What sort of tests should be written for this (if any)

### CODE REVIEW

**Comment by rrousselGit** [comment]
We'd want a test where a package depends on a lint using `dependencies:` and verify that the lints are correctly run

There are similar tests in the packages/custom_lint/test folder

**Comment by dickermoshe** [comment]
Seems to be a bug in how the workspace pubspec.yaml file is written.
I hope to get that working soon...

**Comment by rrousselGit** [comment]
It may not be your fault. The master CI is failing AFAIK

**Comment by dickermoshe** [comment]
@rrousselGit 
It seems the fake project that the tests setup have issues resolving packages:
```console
Building package executable... 
Built test:test.
00:35 +0 -1: With ANSI: true and format: default CLI lists warnings from all plugins and set exit code installAsDevDependency: false [E]                                                                                                                                                                
  Expected: should be done
    Actual: <Instance of '_MapStream<List<int>, String>'>
     Which: emitted • Error: Couldn't resolve the package 'custom_lint_builder' in 'package:custom_lint_builder/src/channel.dart'.
                      Error: Couldn't resolve the package 'test_lint' in 'package:test_lint/test_lint.dart'.
                      Error: Couldn't resolve the package 'test_lint2' in 'package:test_lint2/test_lint2.dart'.
                      lib/custom_lint_client.dart:3:8: Error: Not found: 'package:custom_lint_builder/src/channel.dart'
                      import 'package:custom_lint_builder/src/channel.dart';
                             ^
                      lib/custom_lint_client.dart:4:8: Error: Not found: 'package:test_lint/test_lint.dart'
                      import 'package:test_lint/test_lint.dart' as test_lint;
                             ^
                      lib/custom_lint_client.dart:5:8: Error: Not found: 'package:test_lint2/test_lint2.dart'
                      import 'package:test_lint2/test_lint2.dart' as test_lint2;
                             ^
                      lib/custom_lint_client.dart:17:29: Error: Undefined name 'createPlugin'.
                          {'test_lint': test_lint.createPlugin,
                                                  ^^^^^^^^^^^^
                      lib/custom_lint_client.dart:18:26: Error: Undefined name 'createPlugin'.
                      'test_lint2': test_lint2.createPlugin,
                                               ^^^^^^^^^^^^
                      lib/custom_lint_client.dart:12:3: Error: Method not found: 'runSocket'.
                        runSocket(
                        ^^^^^^^^^
                      
                      
                    • 
                      Failed to start plugins
                      The request analysis.setContextRoots failed with the following error:
                      RequestErrorCode.PLUGIN_ERROR
                      Bad state: Failed to start the plugins.
                      at:
                      #0      SocketCustomLintServerToClientChannel._checkInitializationFail.<anonymous closure> (package:custom_lint/src/v2/server_to_client_channel.dart:223:11)
                      <asynchronous suspension>
                      #1      Future.any.onValue (dart:async/future.dart:628:5)
                      <asynchronous suspension>
  
  package:matcher            expect
  test/cli_test.dart 207:13  main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                 _CustomZone.registerUnaryCallback
  test/cli_test.dart 194:29  main.<fn>.<fn>
  
  Expected: completes to a value that (a string starting with 'Analyzing...' and a string ending with '  lib/another.dart:1:6 • Hello world • hello_world • INFO\n'
              '  lib/another.dart:1:6 • Oy • oy • INFO\n'
              '  lib/main.dart:1:6 • Hello world • hello_world • INFO\n'
              '  lib/main.dart:1:6 • Oy • oy • INFO\n'
              '\n'
              '4 issues found.\n'
              '')
    Actual: <Instance of 'Future<String>'>
     Which: emitted ''
  
  package:matcher            expect
  test/cli_test.dart 217:15  main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                 _CustomZone.registerUnaryCallback
  test/cli_test.dart 194:29  main.<fn>.<fn>
  

To run this test again: dart test test/cli_test.dart -p vm --plain-name 'With ANSI: true and format: default CLI lists warnings from all plugins and set exit code installAsDevDependency: false'
00:35 +0 -1: Some tests failed.   
```

**Comment by dickermoshe** [comment]
It strange. When I manually create a project structure that the tests do, I get an error when resolving deps.
This happens whether it's in dependencies or dev_dependencies
```
Because every version of test_lint from path depends on custom_lint_builder from path which depends on custom_lint 0.6.7, every version of test_lint from path requires custom_lint from hosted.
So, because test_app depends on both custom_lint from path and test_lint from path, version solving failed.
```

**Comment by dickermoshe** [comment]
@rrousselGit 
I pushed the latest test.

If you could take a look at why this doesn't work.
```
dart test test/cli_test.dart -p vm --plain-name 'With ANSI: true and format: default CLI lists warnings from all plugins and set exit code installAsDevDependency: false' --chain-stack-traces
```
But this does
```
dart test test/cli_test.dart -p vm --plain-name 'With ANSI: true and format: default CLI lists warnings from all plugins and set exit code installAsDevDependency: true' --chain-stack-traces
```

**Comment by dickermoshe** [comment]
@rrousselGit 
I spent WAY too much time on this today.

I've written 2 new tests for this PR. And everything passes.

Except..
On `main`, there is 1 test that always fails (`CustomLintWorkspace resolvePluginHost runs offline if possible`) and another that is only fails sometimes (`hot-reload Supports starting custom_lint twice in watch mode at once`).

This is not my fault as you stated.


<details><summary>Logs</summary>
<p>

```
mick@MD-LAPTOP-1:~/dart_custom_lint/packages/custom_lint$ dart test
Building package executable... (2.0s)
Built test:test.
00:08 +18 -1: test/src/workspace_test.dart: CustomLintWorkspace resolvePluginHost runs offline if possible [E]                         
  Exception: Failed to run "pub get" in the client project:
  Resolving dependencies...
  
  pubspec.yaml has no lower-bound SDK constraint.
  You should edit pubspec.yaml to contain an SDK constraint:
  
  environment:
    sdk: '^3.5.0'
  
  See https://dart.dev/go/sdk-constraint
  
  package:custom_lint/src/workspace.dart 831:7  CustomLintWorkspace.runPubGet
  ===== asynchronous gap ===========================
  package:custom_lint/src/workspace.dart 815:5  CustomLintWorkspace.resolvePluginHost
  ===== asynchronous gap ===========================
  test/src/workspace_test.dart 891:9            main.<fn>.<fn>.<fn>
  

To run this test again: /home/mick/development/flutter/bin/cache/dart-sdk/bin/dart test test/src/workspace_test.dart -p vm --plain-name 'CustomLintWorkspace resolvePluginHost runs offline if possible'
01:36 +105 -2: test/server_test.dart: hot-reload Supports starting custom_lint twice in watch mode at once [E]                         
  Bad state: Cannot add event after closing
  dart:async                                                          _StreamController.add
  test/mock_fs.dart 100:17                                            _StdoutOverride.writeln
  package:custom_lint/src/plugin_delegate.dart 232:12                 CommandCustomLintDelegate.serverError
  package:custom_lint/src/v2/custom_lint_analyzer_plugin.dart 233:18  CustomLintServer.handleUncaughtError.<fn>
  ===== asynchronous gap ===========================
  package:custom_lint/src/async_operation.dart 22:14                  PendingOperation.run
  

To run this test again: /home/mick/development/flutter/bin/cache/dart-sdk/bin/dart test test/server_test.dart -p vm --plain-name 'hot-reload Supports starting custom_lint twice in watch mode at once'
  Bad state: Cannot add event after closing
  dart:async                                                          _StreamController.add
  test/mock_fs.dart 100:17                                            _StdoutOverride.writeln
  package:custom_lint/src/plugin_delegate.dart 232:12                 CommandCustomLintDelegate.serverError
  package:custom_lint/src/v2/custom_lint_analyzer_plugin.dart 233:18  CustomLintServer.handleUncaughtError.<fn>
  ===== asynchronous gap ===========================
  package:custom_lint/src/async_operation.dart 22:14                  PendingOperation.run
  
  Bad state: Cannot add event after closing
  dart:async                                                          _StreamController.add
  test/mock_fs.dart 100:17                                            _StdoutOverride.writeln
  package:custom_lint/src/plugin_delegate.dart 232:12                 CommandCustomLintDelegate.serverError
  package:custom_lint/src/v2/custom_lint_analyzer_plugin.dart 233:18  CustomLintServer.handleUncaughtError.<fn>
  ===== asynchronous gap ===========================
  package:custom_lint/src/async_operation.dart 22:14                  PendingOperation.run
  
04:48 +131 -2: Some tests failed.                                                                                                      

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
```

</p>
</details> 


Either way, this PR is done and ready for review!

**Comment by dickermoshe** [comment]
@rrousselGit 

When do you think you guys will be releasing a new version of custom_lint once this PR is merged?

**Comment by dickermoshe** [comment]
I really find this package useful so I decided to spend some time trying to fix this existing bug. 

I managed to fixed the Dart SDK error in the `CustomLintWorkspace resolvePluginHost runs offline if possible` test.
But it turns out there is another bug in the test.  **Yay?!**

I'm not sure what this test is trying to do , so I can't help fix it.
I was under the impression that `dart pub get` needs an internet connection, so I'm not sure what the purpose of this test is, ot how to proceed on fixing it.

Just to be clear, this is your bug, not mine. So IMHO, this shouldn't hold up the PR

**Comment by rrousselGit** [comment]
Hello!
Do you think you could sync with the master branch?

Then, if the CI is passing I'll evaluate whether this can be merged :)

**Comment by dickermoshe** [comment]
I thought you forgot about me 😢

I synced with main and tests pass on my end.

Lmk what changes you would want

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [inline packages/custom_lint/lib/src/workspace.dart:656]: What is the purpose of this?

Comment (rrousselGit) [reply packages/custom_lint/lib/src/workspace.dart:656]: I don't think it's right to pick one or the other. Both can have different constraints
We should respect both constraints, if specified in both dependencies and dev_dependencies.

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [inline packages/custom_lint/lib/src/workspace.dart:685]: That's not right.

Comment (rrousselGit) [reply packages/custom_lint/lib/src/workspace.dart:685]: Actually, it seems like that's how I handled it before....
It shoudn't be though :shrug:

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [inline packages/custom_lint/lib/src/workspace.dart:673]: We should preserve the source of a dependency.
If it's from dependencies, we write it as dependencies. If it is a dev_dependency, we write it as so.
And the constraints should match.

File: packages/custom_lint/lib/src/workspace.dart
Comment (dickermoshe) [reply packages/custom_lint/lib/src/workspace.dart:656]: I did it this way due to this comment:

https://github.com/invertase/dart_custom_lint/blob/dcce163a2864bf794f1d39a4744708573ed773ba/packages/custom_lint/lib/src/workspace.dart#L645-L651

It seemed you wanted to avoid writing the dependency twice.

If that doesn't bother you, i'll change it to write both.

File: packages/custom_lint/lib/src/workspace.dart
Comment (dickermoshe) [reply packages/custom_lint/lib/src/workspace.dart:673]: But what if it's in both?

e.g.
```bash
.
├── b
│   └── pubspec.yaml # drift_dev in dev_dependencies
├── c
│   └── pubspec.yaml # drift_dev in dependencies
└── a
    └── pubspec.yaml # Root Package
```
Project `A` depends on package `B` and `C`.
If `drift_dev`is a `dev_dependencies` in package `B` and a regular `dependency` in `C`, how should it be written? 

I spent a while trying to understand what this code is doing. If my assumptions are wrong please enlighten me

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [reply packages/custom_lint/lib/src/workspace.dart:656]: I think that's out of date. To begin with, it's odd that the comment is commented :P

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [reply packages/custom_lint/lib/src/workspace.dart:673]: Actually, I think we should always be writing plugins to `dependencies:` , not `dev_dependencies:`

Because plugins use them directly rather than through command. If the linter was running against its own source, we'd probably see a warning that plugins should be in "dependencies" instead of dev_dependencies.

It works regardless because Dart sometimes doesn't care. But the current code is probably not right.

Comment (dickermoshe) [reply packages/custom_lint/lib/src/workspace.dart]: >Because plugins use them directly rather than through command. If the linter was running against its own source, we'd probably see a warning that plugins should be in "dependencies" instead of dev_dependencies. 

Good point.

I've updated the code to combine `dev_dependencies` and `dependencies` and to always write them as regular `dependencies` in the generated pubspec. I've updated the tests for this too.

Comment (dickermoshe) [reply packages/custom_lint/lib/src/workspace.dart]: I've also added a test for this specific scenario.

Comment (rrousselGit) [reply packages/custom_lint/lib/src/workspace.dart]: Sorry, let me have another look.

File: packages/custom_lint/lib/src/workspace.dart
Comment (rrousselGit) [inline packages/custom_lint/lib/src/workspace.dart:626]: Nit but we should be able to do `.expand(...)` directly instead of `.map().expand()`

**Review by rrousselGit — APPROVED** [review]
Assuming the CI passes, it's looking good!

**Pull request merged**
